File: recursion/python/binary_search.py
# binary search algorithm

def binary_search(array, target, low=0, high=None):
    """
        Return True if target is found in indicated portion
        of a Python list.
    """

    # if high is not explicitly given, set it to the end of the array.
    if high == None:
        high = len(array) - 1

    if low > high:
        return False  # interval is empty; no match
    else:
        mid = (low + high) // 2
        if target == array[mid]:  # found a match
            return True
        elif target < array[mid]:
            return binary_search(array, target, low, mid - 1)
        else:
            return binary_search(array, target, mid + 1, high)


if __name__ == '__main__':
    array = [1, 23, 55, 92, 84]
    print(binary_search(array, 92))
    print(binary_search(array, 32))

File: recursion/javascript/binarySearch.js
function binarySearch(array, value, low, high) {
    // if low and high not explicitly set, add values
    if (!low) {
        low = 0;
    }
    if (!high) {
        high = array.length - 1;
    }

    // check array is not empty
    if (low > high) {
        return false;
    } else {
        const mid = Math.floor((low + high) / 2);

        if (value === array[mid]) {
            return true; // value found
        } else if (value < array[mid]) {
            return binarySearch(array, value, low, mid - 1);
        } else {
            return binarySearch(array, value, mid + 1, high);
        }
    }
}

const array = [12, 32, 33, 35, 77, 89, 121, 323, 12312];

console.log(binarySearch(array, 33));
console.log(binarySearch(array, 34));
